File: voan/flow.py
"""Information-flow monitor — the confidentiality half of FIDES-style labels.

The 2026 deterministic class (FIDES, CaMeL) tracks data provenance with
confidentiality + integrity labels. Voan's `taint` covers INTEGRITY (a side effect
whose *target* came from untrusted data). This covers CONFIDENTIALITY the other
way: data that came from a SENSITIVE read must not leave through an external sink,
whatever the destination. Declare the tools that return confidential data; the
monitor records the values they returned and flags any external side effect whose
arguments carry one of those values — deterministic data-exfiltration detection,
finer than the session-level Rule of Two (it checks the actual payload, not just
capability counts) and broader than the credential regex (any confidential value,
not only key-shaped strings).

Full per-value flow tracking needs a capability interpreter around the agent
(CaMeL); this is the pragmatic drop-in approximation and is opt-in.
"""
import json
import re

from .taint import is_side_effect

_TOKEN = re.compile(r"[A-Za-z0-9@._+\-]{4,}")
_STOP = {"true", "false", "null", "none", "http", "https", "message", "status",
         "amount", "date", "name", "type", "value", "sent", "from", "with", "this"}


def _tokens(output):
    text = output if isinstance(output, str) else json.dumps(output, default=str)
    return {t.lower() for t in _TOKEN.findall(text) if t.lower() not in _STOP}


class FlowMonitor:
    def __init__(self, confidential_tools=()):
        # tools whose OUTPUT is confidential (balances, PII, secrets, private docs)
        self.conf_tools = set(confidential_tools)
        self.secrets = set()

    def reset(self):
        self.secrets = set()
        return self

    def observe(self, tool, output):
        if tool in self.conf_tools:
            self.secrets |= _tokens(output)

    def leaks(self, action):
        """Return the first confidential value carried by an external side effect's
        arguments, else None. Reads never leak (they don't send anything out)."""
        if not is_side_effect(action.tool) or not self.secrets:
            return None
        blob = json.dumps(action.args if isinstance(action.args, dict) else {},
                          default=str).lower()
        for tok in _tokens(blob):
            if tok in self.secrets:
                return tok
        return None

File: sdk-js/src/flow.ts
// Information-flow monitor — the confidentiality half of FIDES-style labels. JS port
// of voan/flow.py. Data from a declared confidential-source tool must not leave via
// an external sink, whatever the destination. Deterministic.
import { isSideEffect } from "./auto.ts";
import type { Action } from "./schema.ts";

const TOKEN = /[A-Za-z0-9@._+-]{4,}/g;
const STOP = new Set(["true", "false", "null", "none", "http", "https", "message",
  "status", "amount", "date", "name", "type", "value", "sent", "from", "with", "this"]);

function tokens(output: unknown): Set<string> {
  const text = typeof output === "string" ? output : JSON.stringify(output);
  return new Set((text.match(TOKEN) ?? []).map((t) => t.toLowerCase())
    .filter((t) => !STOP.has(t) && t.length >= 4));
}

export class FlowMonitor {
  private confTools: Set<string>;
  private secrets = new Set<string>();

  constructor(confidentialTools: Iterable<string> = []) {
    this.confTools = new Set(confidentialTools);
  }

  reset(): this { this.secrets = new Set(); return this; }

  observe(tool: string, output: unknown): void {
    if (this.confTools.has(tool)) for (const t of tokens(output)) this.secrets.add(t);
  }

  /** First confidential value carried by an external side-effect's args, else null. */
  leaks(action: Action): string | null {
    if (!isSideEffect(action.tool) || !this.secrets.size) return null;
    const args = (action.args && typeof action.args === "object") ? action.args : {};
    for (const tok of tokens(args)) if (this.secrets.has(tok)) return tok;
    return null;
  }
}
